### ISSUE TITLE
Unclassified segments when doing simple clustering?

### ISSUE THREAD

**Issue by gabrielparriaux** [juba/rainette#34]
Hello @juba, 

I thought that it was only when doing double clustering that there was this option not to force the classification of some segments and that some segments got a NA value as classification. 

But I was surprised to do a simple clustering, get back the clusters with `cutree_rainette()` and have some of them being NA.

Did I do something wrong? Is it normal that some segments are not classified when performing a simple rainette clustering?

Thanks a lot for your opinion about that and all the best!

Gabriel

**Comment by juba** [comment]
In general when a document gets NA as cluster in a simple classification it doesn't have any content (it is very short and/or only consisted of terms that have been filtered out).

**Comment by gabrielparriaux** [comment]
Oh, ok! So, it must be the reason… I will check the content of those segments!

**Comment by gabrielparriaux** [comment]
Hello @juba,

Sorry to reopen this, but I would like to come back to this question.

In an analysis that I perform, I have quite a lot of small segments because they are coming from teachers classroom discourse and they often take very short utterances.

Before performing clustering, I remove from the dfm all segments with less than 3 tokens and all words present in less than 3 documents.

My dfm contains 306 segments and 173 unique tokens.

I use `rainette()` with the following parameters:

```r
rainette(my_dfm, k = 8, min_segment_size = 10, min_split_members = 8)
```

As a result, only one segment is not classified in any cluster. It has size = 3. 

A lot of other segments are also the same size.

In the case of the only unclassified segment, it has no other segment coming from the same source document, so it can not be merged with a following or previous one. But it is also the case of a lot of other segments with the same size of 3, which have no other segments coming from the same source document, that could not be merged with a following or previous one, but that were classified in a cluster anyway.

Is there a specific reason why this segment has not been classified, but others that were the same size got classified?

If I lower the min_segment_size to 3 and use:

```r
rainette(my_dfm, k = 8, min_segment_size = 3, min_split_members = 8)
```

Then all segments are classified in clusters.

Should I always adapt the `min_segment_size` to the smallest size of segments that I have?

Thanks a lot for your advice and all the best,
Gabriel

**Comment by juba** [comment]
Hi,

No you shouldn't have to set min_segment_size to the size of the smallest segment in your corpus, as this would prevent any merging  between segments from the same document when this is possible.

I don't really understand why one segment would be classified when min_segment_size is 3 but not when it is 10. Do you have any warnings in the `rainette()` output such as "some segments will have a size < min_segment_size" or "some documents don't have any term, they won't be assigned to any cluster." ?

**Comment by gabrielparriaux** [comment]
Hi, 

Thanks a lot for your opinion about that.

Yes, I get the following warning in the `rainette()` output:

```r
Warning message:
In rainette::merge_segments(dtm, min_segment_size = min_segment_size,  :
  some segments will have a size < min_segment_size
```

**Comment by juba** [comment]
You get this warning even with min_segment_size=3 or only when min_segment_size=8 ?
And you don't get any "some documents don't have any term, they won't be assigned to any cluster." ?

**Comment by gabrielparriaux** [comment]
I get the warning in both cases: with `min_segment_size = 3` and `min_segment_size = 8`, the warning is exactly the same as displayed just in my previous message.

I don’t get the other warning you mention ("some documents don't have any term, they won't be assigned to any cluster.").

**Comment by juba** [comment]
Could you run the following code and copy paste the output ?

```r
print(min(rowSums(my_dfm)))
my_dfm2 <- quanteda::dfm_remove(my_dfm, "")
print(min(rowSums(my_dfm2)))
```

**Comment by gabrielparriaux** [comment]
Voilà l’output:
```r
[1] 3
[1] 3
```

**Comment by juba** [comment]
And what about:

```r
print(min(rowSums(my_dfm)))
my_dfm2 <- quanteda::dfm_remove(my_dfm, "")
print(min(rowSums(my_dfm2)))
my_dfm3 <- quanteda::dfm_weight(my_dfm2, scheme = "boolean")
print(min(rowSums(my_dfm3)))
```

I realize it may be faster if you could just extract the content of the unclassified segment 😅

**Comment by gabrielparriaux** [comment]
Output is:
```r
[1] 3
[1] 3
[1] 2
```

And the content of the unclassified segment is this one:
```r
Tokens consisting of 1 document and 107 docvars.
cl02_pr1_text374_1 :
[1] "tu"       "rappeler" "effacer" 
```

**Comment by juba** [comment]
Ok, so the warning is about a segment of size 2 because it should have the same term several times, but this is not the one which is not classified so this is not the issue.

Unfortunately I don't really understand why this segment specifically is not classified, to do that I would have to have access to the data and try to analyze what the clustering is doing at each step...

**Comment by gabrielparriaux** [comment]
Thank you for your analysis and help, that’s really appreciated.

Should I email you .rds files with my dfm and corpus? I’m afraid to waste too much of your time with my questions…

**Comment by juba** [comment]
Yes, you can send me your data if you want, that would be a way to check if the behavior is normal or if there is a bug somewhere.
